Filtering skeleton

CremaLuca opened the pull request:
## Premise
This PR aims to implement partially what has been discussed about this [Trello board](https://trello.com/c/TY4yCJPi/54-filtering-pattern)

## Features

Some untold rules:

- When a stream is flagged as `is_finished` no atoms shall be added to it.
- A filter is finished when all of its outputs are flagged as `is_finished`

#### feat(`stream.py`)
It's a list of atoms that uses a `StreamIterator` as default iterator.
**+** StreamIterator has a `has_next()` method and when calling `next()` it returns an already removed atom from the stream.

#### feat(`filter.py`)
It's an abstract class for filters, defines the inputs and the outputs of the filter, the parameters of the `execute()` method and the `is_finished()` method.

#### feat(`filter_layer.py`)
Defines a set of filters, that's it, nothing more.

#### feat(`filter_list.py`)
It's the sequence of filter layers. The class defines the behaviour of execution of the filters.

#### feat(`utils/time_handler.py`)
Utils class to avoid importing datetime everytime we need to work on datetimes, should be updated on later PR's to include operations done in the `downloader` classes.

#### test(`utils/time_handler.py`)
**+** Test for `str_to_datetime`
**+** Test for `datetime_to_str`

CremaLuca committed:
feat: starting off filtering

CremaLuca committed:
feat:
~ cleaned filter_layer, now extends list
+ filter_list: moved source_filter addition to its own method, manages multiple source streams, has a callback for when atoms are produced
+ added Stream, a collection that returns an iterator that has has_next() method

CremaLuca committed:
documentation(filter,py)
+ added get_output_stream(index) method

ricdezen commented on otri/filtering/stream.py:
```diff
@@ -0,0 +1,37 @@
+        Pops the first element of the given collection.
+        '''
+        value = self.iterable[0]
+        del self.iterable[0]
```
Funge? Toglierlo lo toglie, ma lo cancella anche dalla memoria penso (???)

CremaLuca replied:
È quello che vogliamo, vogliamo liberare la memoria dello stream attuale una volta che è stato letto un atomo, se ne occuperà il filtro a ficcare l'atomo in un altro stream per tenerlo vivo (o distruggerlo perché non serve)

ricdezen replied:
Il punto è che così distruggi secco l'atomo, il che assume che il ricevente ne abbia fatto una deep copy, è necessario? La garbage collection lo rimuoverebbe ugualmente se non fosse più usato

CremaLuca replied:
Il garbage collector non lo rimuoverebbe finché la lista non viene più usata, se dobbiamo processare 1TB di atomi e abbiamo tre filtri alla fine ci troveremo con 3TB di stream che contengono i tre stadi di ogni atomo, invece vogliamo che quel 1TB rimanga uno solo

ricdezen replied:
Ok provando ho scoperto che un del su un elemento della lista lo rimuove solo, ero convinto che avrebbe cancellato anche l'oggetto. Non lo fa perchè gli elementi di una lista sono oggetti a parte, dei puntatori.

CremaLuca committed:
feat: moved is_finished to streams
+ used get_output_stream method in filter_list and successive_source_filter
+ fixed doc in filter_layer
+ started interpolation filter

CremaLuca committed:
feat: interpolation_filter.py

ricdezen commented on otri/filtering/filters/interpolation_filter.py:
```diff
@@ -0,0 +1,119 @@
+                "The atom list is not datetime-ordered or two atoms have the same datetime:\n{}\n{}".format(atom1, atom2))
+        return value
+
+    def __timedelta_from_interval(self, interval: str):
```
A dict sounds easier

CremaLuca committed:
fix: interpolation_filter now uses right progress

CremaLuca committed:
style: interpolation_filter

CremaLuca committed:
refactor: interpolation_filter uses a dict to define supported intervals

CremaLuca committed:
style(interpolation_filter): reduced line sizes by using time_handler

CremaLuca committed:
fix: wrong time_handler import

CremaLuca committed:
feat: better test.py using real data

CremaLuca committed:
refactor(filter): moved is_finished inside
~ Now filter_list checks every last layer's output stream to call the listener

CremaLuca committed:
style: fixed some codacy issues

CremaLuca committed:
style: fixed some other codacy issues

CremaLuca committed:
style: fixed some more codacy issues

CremaLuca committed:
feat(stream.py): added check on append and insert if the stream is already closed

CremaLuca committed:
test(stream.py) added some stream tests

Commit:
fix (database_stream.py): added missing return statement

Commit:
Merge branch 'feat/database-stream' of https://github.com/OTRI-Unipd/OTRI into feat/database-stream

CremaLuca committed:
fix(stream): wrong is_finished call
+ test(stream)

CremaLuca committed:
test(stream): further testing
~ changed stream.is_finished in stream.is_closed
+ added exception  if stream is already closed when closing it again

CremaLuca committed:
Merge branch 'develop' into feat/filtering

# Conflicts:
#	test.py

CremaLuca committed:
fix(filter): output streams were initialised with the same object for every list element.
+ test(filter_test)

CremaLuca committed:
style(filter): renamed private attributes
+ test(filter)

CremaLuca committed:
feat(merge_filter): removed successive source filter
- successive_source_filter, it was useless
~ filter_list now doesn't need a source filter

CremaLuca committed:
feat(stream): is_closed can now be defined in the constructor

CremaLuca committed:
style(filter_list): fixed some codacy issues

Commit:
fix (filter.py): missing type hint
feat (nuplicator_filter.py):
+ added NUplicatorFilter class, makes N copies of a single input stream and appends them to N output streams

CremaLuca committed:
test(merge_filter)

Commit:
feat (nuplicator_filter_test.py) : simple tests for a simple class

CremaLuca committed:
feat(filter_list): started writing more complex algorithm for execution

CremaLuca committed:
feat(multiplier_filter)

CremaLuca committed:
feat(summer_filter)

Commit:
feat (average_filter_test.py):
Added tests for avg/sum calculating filter

Commit:
feat (average_filter.py):
+ AverageFilter class for calculating the average and sum of a set of keys

CremaLuca committed:
feat(summer_filter.py): a different constant for every key

ricdezen commented on otri/filtering/filters/summer_filter.py:
```diff
@@ -10,21 +10,21 @@ class SummerFilter(Filter):
     '''
 
-    def __init__(self, input_stream: Stream, keys_to_change: Collection[str], const : int):
+    def __init__(self, input_stream: Stream, keys_to_change: Collection[str], constants : Mapping):
```
Can directly be a mapping of the keys to their constants

CremaLuca replied:
Sure

CremaLuca committed:
stlye(summer_filter): merged keys to change and constants

Commit:
fix (nuplicator_filter.py): changed default value for deep_copy init param to True

Commit:
fix (nuplicator_filter_test.py): fixed test

ricdezen committed:
Merge pull request #14 from OTRI-Unipd/feat/n-uplicator-filter

N-Uplicator Filter

Commit:
Merge branch 'feat/filtering' of https://github.com/OTRI-Unipd/OTRI into feat/avg-filter

Commit:
fix (average_filter.py):
+ added check to avoid closing the output stream twice
+ Added class specs

fix (nuplicator_filter.py):
+ added check to avoid closing the output stream twice

ricdezen committed:
Merge pull request #13 from OTRI-Unipd/feat/avg-filter

Average Filter

CremaLuca committed:
feat(filter_list): added possibility to instantiate list with layer array

CremaLuca committed:
Merge branch 'feat/filtering' of https://github.com/OTRI-Unipd/OTRI into feat/filtering

CremaLuca committed:
feat(test.py): failed testing, average is always 0

Commit:
fix(test.py): moved print statement after execution

Commit:
feat (test.py): added multiplier filter to test

CremaLuca committed:
style(filter_list): removed useless method get_output_streams

Commit:
Merge branch 'feat/filtering' of https://github.com/OTRI-Unipd/OTRI into feat/database-stream

Commit:
fix (profile_batch.py): changed to __next__ calls instead of for loop

fix (database_stream.py): changed DatabaseStream and Iterator to better match standard Stream structure
+ Added internal buffer to PostgreSQL iterator for better has_next calls

Commit:
feat (database_stream.py) :
+ added reference to parent strem to iterator, so that it can be closed when the stream is over
+ added check for cursor status on __next__ iterator method

Commit:
fix (database_stream.py): various minor specs updates

CremaLuca committed:
feat(math_filter) edited summer filter to sustain any single operation on items values

Commit:
fix (database_stream.py) : OCD

Commit:
fix (nuplicator_filter.py) : made unnecessarily public fields into private

Commit:
feat (generic_filter.py):
+ Added class GenericFilter that applies some Callable to a Stream

test (generic_filter_test.py) : basic tests

CremaLuca committed:
Merge pull request #16 from OTRI-Unipd/hotfix/db-stream-next

Hotfix/db stream next

CremaLuca committed:
feat(autocorrelation):
+ takes data from DB
+ new generic filter that converts from tuple to atom

CremaLuca committed:
Merge pull request #17 from OTRI-Unipd/feat/math-filter

Math filter

CremaLuca committed:
feat(autocorrelation):
+ summer filter became math filter

CremaLuca committed:
fix(autocorrelation): some wrong lines

CremaLuca committed:
fix(autocorrelation): removed debug print

CremaLuca committed:
feat(autocorrelation): calculation for every ticker

Commit:
feat (statistics_filter.py):
- Removed AverageFilter
+ Added StatisticsFilter, builder pattern Filter allowing to compute various statistics on the input Stream

test (statistics_filter_test.py):
+ All the old tests for the AverageFilter, plus a test for the minimum value and the maximum value

Commit:
fix (statistics_filter.py):
+ StatisticsFilter, added specs to methods

Commit:
fix (statistics_filter.py):
+ StatisticsFilter, added missing specs to __init__ method

CremaLuca committed:
Merge pull request #18 from OTRI-Unipd/feat/statistics-filter

Statistics Filter

CremaLuca committed:
Merge branch 'feat/filtering' of https://github.com/OTRI-Unipd/OTRI into feat/filtering

CremaLuca committed:
feat(autocorrelation): now calculates autocorr for every russell3000 ticker

Commit:
fix (statistics_filter.py):
+ StatisticsFilter.__init__ : spec fix

Commit:
test (generic_filter_test.py):
+ added tests for empty input stream
+ added tests for repeated calls after closing output

test (nuplicator_filter_test.py):
+ added tests for empty input stream
+ added tests for repeated calls after closing output

test (statistics_filter_test.py):
+ added tests for empty input stream
+ added tests for repeated calls after closing output

Commit:
feat (multiplier_filter.py):
+ Renamed to phase_filter.py

feat (phase_filter.py):
+ PhaseFilter, generic filter class applying some operation to the keys in the input stream and putting the result in a single output stream. Made from minor touchups to MultiplierFilter
+ PhaseMulFilter, subclass of PhaseFilter directly providing multiplication lambdas to the keys
+ PhaseDeltaFilter, subclass of PhaseFilter directly providing delta lambdas to the keys

test (phase_filter_test.py):
+ tests for PhaseFilter
+ tests for PhaseMulFilter
+ tests for PhaseDeltaFilter

CremaLuca committed:
Merge pull request #19 from OTRI-Unipd/hotfix/delta-filter

Delta Filter and Multiplier Refactor

CremaLuca committed:
feat(delta_autocorrelation):
+ Moved autocorrelation calculation inside a function
+ Added autocorrelation_delta function that uses deltas
~ Temporarily removed normalization filter

CremaLuca committed:
feat:
+ parametrized distance
+ calculation of time took by both metods

CremaLuca committed:
feat(stream): changed thrown exception in next to StopIter so that it can be placed inside a for loop

CremaLuca committed:
fix(autocorrelation): removed matplotlib

CremaLuca committed:
doc(autocorrelation): added module documentation

CremaLuca commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 7
           

Complexity increasing per file
==============================
- otri/filtering/filters/math_filter.py  5
- otri/filtering/filters/generic_filter.py  4
- otri/filtering/filters/merge_filter.py  6
- test.py  1
- autocorrelation.py  3
- test/filtering/filter_test.py  2
- test/filtering/filters/nuplicator_filter_test.py  2
- otri/filtering/filters/phase_filter.py  6
- otri/utils/time_handler.py  1
- test/filtering/filters/generic_filter_test.py  3
- test/utils/time_handler_test.py  1
- otri/filtering/filters/interpolation_filter.py  6
- otri/filtering/filter.py  3
- otri/filtering/filter_list.py  13
- test/filtering/filters/merge_filter_test.py  1
- otri/database/database_stream.py  1
- otri/filtering/filters/nuplicator_filter.py  6
- otri/filtering/filter_layer.py  3
- otri/filtering/stream.py  2
- test/filtering/filters/phase_filter_test.py  2
- otri/filtering/filters/statistics_filter.py  6
- test/filtering/stream_test.py  1
- test/filtering/filters/statistics_filter_test.py  2
         

Clones added
============
- autocorrelation.py  2
- test/filtering/filters/merge_filter_test.py  2
- test/filtering/filters/phase_filter_test.py  5
- test/filtering/filters/statistics_filter_test.py  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939&bid=17990617)

CremaLuca commented on otri/filtering/filters/generic_filter.py:
```diff
@@ -0,0 +1,42 @@
+from ..filter import Filter, Stream, Collection
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused Collection imported from filter](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939)

CremaLuca replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused Collection imported from filter](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939)

CremaLuca commented on otri/filtering/filters/statistics_filter.py:
```diff
@@ -0,0 +1,167 @@
+from ..filter import Filter, Stream, Collection
+from typing import Sequence, Mapping, Dict
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused Dict imported from typing](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939)

CremaLuca replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused Dict imported from typing](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939)

CremaLuca commented on test.py:
```diff
@@ -0,0 +1,86 @@
+from otri.filtering.filters.math_filter import MathFilter
+from otri.filtering.filters.average_filter import AverageFilter
+from otri.downloader.yahoo_downloader import YahooDownloader, date
+from otri.downloader.alphavantage_downloader import AVDownloader
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused AVDownloader imported from otri.downloader.alphavantage_downloader](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939)

CremaLuca replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused AVDownloader imported from otri.downloader.alphavantage_downloader](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939)

CremaLuca commented on otri/filtering/filters/nuplicator_filter.py:
```diff
@@ -0,0 +1,47 @@
+from ..filter import Filter, Stream, Collection
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused Collection imported from filter](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939)

CremaLuca replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused Collection imported from filter](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939)

CremaLuca commented on otri/filtering/filters/statistics_filter.py:
```diff
@@ -0,0 +1,167 @@
+from ..filter import Filter, Stream, Collection
+from typing import Sequence, Mapping, Dict
+from numbers import Number
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused Number imported from numbers](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939)

CremaLuca replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused Number imported from numbers](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939)

CremaLuca commented on autocorrelation.py:
```diff
@@ -0,0 +1,206 @@
+        keys_operations={k: lambda value: value/v
+                         for k, v in stats_filter.get_max().items()}
+    )
+    f_layer_normalize = FilterLayer([normalize_filter])
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused variable 'f_layer_normalize'](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939)

CremaLuca replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused variable 'f_layer_normalize'](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939)

CremaLuca commented on otri/filtering/filters/math_filter.py:
```diff
@@ -0,0 +1,42 @@
+from ..filter import Filter, Stream, Collection
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused Collection imported from filter](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939)

CremaLuca replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused Collection imported from filter](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939)

CremaLuca committed:
feat(filter_styles.py): example of styles that we came up to improve current style

CremaLuca commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           

Complexity increasing per file
==============================
- otri/filtering/filter_layer.py  3
- otri/filtering/filters/math_filter.py  5
- otri/utils/time_handler.py  1
- test/filtering/filter_test.py  2
- test.py  1
- test/filtering/filters/generic_filter_test.py  3
- test/filtering/filters/nuplicator_filter_test.py  2
- otri/filtering/filter.py  3
- autocorrelation.py  3
- otri/filtering/filters/nuplicator_filter.py  6
- test/filtering/filters/statistics_filter_test.py  2
- test/filtering/filters/phase_filter_test.py  2
- otri/filtering/filters/statistics_filter.py  6
- otri/filtering/filters/phase_filter.py  6
- test/filtering/stream_test.py  1
- test/utils/time_handler_test.py  1
- otri/filtering/filters/merge_filter.py  6
- otri/filtering/filter_list.py  13
- otri/filtering/filters/interpolation_filter.py  6
- otri/filtering/filters/generic_filter.py  4
- otri/database/database_stream.py  1
- test/filtering/filters/merge_filter_test.py  1
- otri/filtering/stream.py  2
         

Clones added
============
- filter_styles.py  2
- autocorrelation.py  2
- test/filtering/filters/statistics_filter_test.py  7
- test/filtering/filters/phase_filter_test.py  5
- test/filtering/filters/merge_filter_test.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939&bid=17990617)

CremaLuca commented on filter_styles.py:
```diff
@@ -0,0 +1,170 @@
+                PhaseDeltaFilter(
+                    input="ticker1_interp",
+                    output="ticker1_delta"
+                    keys=KEYS_TO_CHANGE,
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Syntax Error (MC0000)](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939)

ricdezen commented on autocorrelation.py:
```diff
@@ -0,0 +1,206 @@
+Module to calculate autocorrelation within a single ticker.
+"""
+
+__autor__ = "Riccardo De Zen <ricdezern@gmail.com>, Luca Crema <lc.crema@hotmail.com>"
```
```suggestion
__autor__ = "Riccardo De Zen <riccardodezen98@gmail.com>, Luca Crema <lc.crema@hotmail.com>"
```

CremaLuca committed:
fixed de zen's email

Co-authored-by: Riccardo De Zen <56487105+RicDeZen@users.noreply.github.com>
